src/python/pants/backend/python/util_rules/pex_from_targets.py

```diff
@@ -244,13 +244,25 @@ async def pex_from_targets(
         # In requirement strings, Foo_-Bar.BAZ and foo-bar-baz refer to the same project. We let
         # packaging canonicalize for us.
         # See: https://www.python.org/dev/peps/pep-0503/#normalized-names
-        exact_req_projects = {
-            canonicalize_project_name(Requirement.parse(req).project_name) for req in exact_reqs
-        }
+
+        url_reqs = set()  # E.g., 'foobar@ git+https://github.com/foo/bar.git@branch'
+        name_reqs = set()  # E.g., foobar>=1.2.3
+        name_req_projects = set()
+
+        for req_str in exact_reqs:
+            req = Requirement.parse(req_str)
+            if req.url:  # type: ignore[attr-defined]
+                url_reqs.add(req)
+            else:
+                name_reqs.add(req)
+                name_req_projects.add(canonicalize_project_name(req.project_name))
+
         constraint_file_projects = {
             canonicalize_project_name(req.project_name) for req in constraints_file_reqs
         }
-        unconstrained_projects = exact_req_projects - constraint_file_projects
+        # Constraints files must only contain name reqs, not URL reqs (those are already
+        # constrained by their very nature). See https://github.com/pypa/pip/issues/8210.
```

Comment:
Noting that Pip now has support for URL reqs in constraint files via: https://github.com/pypa/pip/pull/9673
Pex won't be using that Pip for a while, but just noting.

https://github.com/pypa/pip/issues/8210 spawned -> https://github.com/pypa/pip/issues/8253 which was fixed by -> https://github.com/pypa/pip/pull/9673